src/GraphMetadataViewExtension/GraphMetadataViewExtension.cs

```diff
@@ -0,0 +1,92 @@
+using Dynamo.Wpf.Extensions;
+using System.Windows;
+
+
+namespace Dynamo.GraphMetadata
+{
+    public class GraphMetadataViewExtension : ViewExtensionBase, IExtensionStorageAccess
+    {
+        internal static string extensionName = "Properties";
+        internal GraphMetadataViewModel viewModel;
+        private GraphMetadataView graphMetadataView;
+        private ViewLoadedParams viewLoadedParamsReference;
+        private MenuItem graphMetadataMenuItem;
+
+        public override string UniqueId => "28992e1d-abb9-417f-8b1b-05e053bee670";
+
+        public override string Name => extensionName;
+
+        public override void Loaded(ViewLoadedParams viewLoadedParams)
+        {
+            if (viewLoadedParams == null) throw new ArgumentNullException(nameof(viewLoadedParams));
+
+            this.viewLoadedParamsReference = viewLoadedParams;
+            this.viewModel = new GraphMetadataViewModel(viewLoadedParams);
+            this.graphMetadataView = new GraphMetadataView();
+            graphMetadataView.DataContext = viewModel;
+
+            // Add a button to Dynamo View menu to manually show the window
+            this.graphMetadataMenuItem = new MenuItem { Header = Resources.MenuItemText, IsCheckable = true };
+            this.graphMetadataMenuItem.Checked += MenuItemCheckHandler;
+            this.graphMetadataMenuItem.Unchecked += MenuItemUnCheckedHandler;
+            this.viewLoadedParamsReference.AddExtensionMenuItem(this.graphMetadataMenuItem);
+        }
+
+        private void MenuItemUnCheckedHandler(object sender, RoutedEventArgs e)
+        {
+            viewLoadedParamsReference.CloseExtensioninInSideBar(this);
+        }
+
+        private void MenuItemCheckHandler(object sender, RoutedEventArgs e)
+        {
+            this.viewLoadedParamsReference?.AddToExtensionsSideBar(this, this.graphMetadataView);
+        }
+
+        #region Storage Access implementation
+        public void OnWorkspaceOpen(Dictionary<string, string> extensionData)
+        {
+            foreach (var kv in extensionData)
+            {
+                if (string.IsNullOrEmpty(kv.Key)) continue;
+
+                var valueModified = kv.Value == null ? string.Empty : kv.Value;
+
+                this.viewModel.AddCustomProperty(kv.Key, valueModified);
+            }
+        }
+
+        public void OnWorkspaceSaving(Dictionary<string, string> extensionData, SaveContext saveContext)
+        {
+            // Clearing the extensionData dictionary before adding new values
+            // as the GraphMetadataViewModel.CustomProperties is the true source of the custom properties
+            extensionData.Clear();
+            foreach (var p in this.viewModel.CustomProperties)
+            {
+                extensionData[p.PropertyName] = p.PropertyValue;
+            }
+        }
+
+
+        #endregion
+
+        protected virtual void Dispose(bool disposing)
+        {
+            viewModel.Dispose();
```

Comment:
Why are we disposing the view model ? In theory a view extension could get unloaded and the viewmodel still stands.
And in the shutdown sequence I imagine viewmodel might still be in use for a while after our extension was disposed or ?

Reply:
> In theory a view extension could get unloaded and the viewmodel still stands.

I dont think this can actually happen? AFAIK the only way to unload an extension is to uninstall the package it comes from and uninstalling a package needs a restart of Dynamo to happen. And if this actually can happen, why would we want the ViewModel of that extension to kept around, i would assume that if you remove an extension everything belonging to that extension should go/

> And in the shutdown sequence I imagine viewmodel might still be in use for a while after our extension was disposed or ?

hm, not sure i see what the viewModel would be in use for? also the dispose on the viewModel only unsubscribe events so in theory everything else is still available.

Reply:
Aaa yes, sorry, I made a confusion, I don't know why I had DynamoViewModel in mind